Src/VimCore/SelectionChangeTracker.fs

```diff
@@ -101,11 +101,38 @@ type internal SelectionChangeTracker
     /// If the caret changes position and it wasn't initiated by VsVim then we should be 
     /// adjusting the screen to account for 'scrolloff'
     member x.OnPositionChanged() = 
-        // Don't update the caret if a selection is occuring.  This could be a user double clicking, 
-        // dragging, etc ...  Don't want to update the caret in that case, let the mode change handle
-        // that.
-        if not _vimBuffer.IsProcessingInput then
-            _commonOperations.EnsureAtCaret ViewFlags.ScrollOffset
+
+        // Don't apply the scroll offset if it isn't applicable, if the
+        // text view is currently being laid out, or if we are in the middle
+        // of processing input. If we are processing input then the mode is
+        // is responsible for ensuring that the scroll offset is obeyed,
+        // so let the mode handle it.
+        if
+            _vimBuffer.GlobalSettings.ScrollOffset > 0
+            && not _textView.InLayout
+            && not _vimBuffer.IsProcessingInput
+        then
+
+            // Do the update being cautious that anything could have
+            // happened between when we posted it and when it actually
+            // runs.
+            let doUpdate () =
+                try
+                    if not _textView.IsClosed then
+                        _commonOperations.EnsureAtCaret ViewFlags.ScrollOffset
+                with
+                | _ -> ()
+
+            // Delay the update to give whoever changed the caret position the
+            // opportunity to scroll the view according to their own needs.
+            // If another extension moves the caret far offscreen and then
+            // centers it if the caret is not onscreen, then reacting too
+            // early to the caret position will defeat their offscreen
+            // handling. An example is double-clicking on a test in an
+            // unopened document in "Test Explorer".
+            let context = System.Threading.SynchronizationContext.Current
+            if context <> null then
+                context.Post((fun _ -> doUpdate()), null)

```

Comment:
What happens if this is done as a part of say macro execution? In that case the macro would completely execute and then all of the posted operations here would execute at the conclusion of the macro.

Reply:
I believe that even if we were to "try to keep the screen up-to-date" during macro execution, WPF would not update the window until control returns back to the event loop. This can be verified by, say, recording a macro for `G1GG1GG1G` that repeatedly goes to the beginning and end of the buffer and observing that the screen doesn't "flash through" all the intermediate updates.

In addition, this operation is supposed to be "idempotent" which is to say that repeating it 'n' times has the same effect as executing it once. Alternatively, it only needs to run "before the next time we are waiting for input from the user."